### ISSUE TITLE
Gradle plugin should not use JUnit 4 Engine when runJunit4 is false

### ISSUE THREAD

**Issue by marcphilipp** [junit-team/junit5#285]
## Status Quo

Behind the scenes, the JUnit Gradle plugin uses the `ConsoleRunner` (from the `junit-console` module) to actually execute the test plan. Due to #146, the `junit-console` includes `junit4-engine`, `junit5-engine`, and `junit4-runner` as mandatory run-time dependencies. Consequently, the `junit4` and `junit5` engines are always on the classpath for the JUnit Gradle plugin, which means that those engines are _included_ by default. The `junit4` engine is therefore _included_ even if the `runJunit4` flag for the plugin has been explicitly set to `false`, which is both misleading and unacceptable.

## Related Issues

- #291
- #295
- #303 
- #304 
- #307

## Deliverables

- [x] Add a direct compile-time dependency on `junit-console` in `junit-gradle.gradle`.
- [x] Refer to the `ConsoleRunner` via a class reference within the JUnit Gradle plugin instead of via a hard-coded string.
- [x] Introduce a new `disableStandardTestTask` configuration flag for disabling execution of the standard Gradle `test` task; default the value of the flag to `true`.
- [x] Remove the `runJunit4` configuration flag since it has been superseded by `engines { include ... }` and `engines { exclude ... }` support.
- [x] Remove obsolete `version` property from plugin configuration.
- [ ] Remove all programmatic dependency management within the JUnit Gradle plugin.
 - Everything has been removed except the dependency on the `junit-console`
 - See discussion in [Gradle Forum](https://discuss.gradle.org/t/how-to-add-plugins-dependencies-to-custom-task/18025)

**Comment by sbrannen** [comment]
I think this is a good opportunity to rename `runJunit4` to `runJUnit4` or `runJUnit4Tests` as well, just in time for M1.

**Comment by sbrannen** [comment]
I actually don't think this feature every worked properly.

I think the only proper way to implement it is to explicitly _exclude_ the "junit4" engine when configuring the `ConsoleRunner` if `runJunit4` is set to `false`. In other words, we need another flag that has the opposite effect of `requiredEngine` -- something like `excludedEngines`.

Similarly, I think `requiredEngine` should be replaced with a list of `requiredEngines`.

**Comment by sbrannen** [comment]
### Team Decisions

- Replace `requiredEngine` with `includedEngines`
- Introduce `excludedEngines`
- Remove `runJunit4` (superseded by `includedEngines`)
- Introduce `disableTestTask` for disabling the standard Gradle `test` task

This have been added as _Deliverables_ in this issue's description.

**Comment by sbrannen** [comment]
Blocked until #291 has been resolved.

**Comment by sbrannen** [comment]
_Deliverables_ have been updated to align with the goals of #291.

**Comment by sbrannen** [comment]
Since #291 has been resolved, this issue is no longer _blocked_.

**Comment by sbrannen** [comment]
This issue is _blocked_ until #295 is resolved.

**Comment by sbrannen** [comment]
This issue is no longer _blocked_.

**Comment by sbrannen** [comment]
FYI: I'm working on the first two deliverables.

**Comment by sbrannen** [comment]
## Team Decisions

- Add a direct compile-time dependency on `junit-console` in `junit-gradle.gradle`.
- Refer to the `ConsoleRunner` via a class reference within the JUnit Gradle plugin instead of via a hard-coded string.
- Remove all programmatic dependency management within the JUnit Gradle plugin.
- Introduce a new `disableStandardTestTask` configuration flag for disabling execution of the standard Gradle `test` task; default the value of the flag to `true`.

The _Deliverables_ have been updated accordingly.

**Comment by sbrannen** [comment]
_in progress_

**Comment by sbrannen** [comment]
***Status***

Regarding removal of all programmatic dependency management within the JUnit Gradle plugin, everything has been removed except the dependency on `junit-console`.

See the question I posted to the [Gradle Forum](https://discuss.gradle.org/t/how-to-add-plugins-dependencies-to-custom-task/18025) for details.

**Comment by sbrannen** [comment]
**Resolution**

Resolved remaining issues in 430e21d6fb3908d0415c553bf38feb2e3f97f0a2:

_Ensure JUnit Gradle plugin does not influence project classpaths_

Prior to this commit, the JUnit Gradle plugin modified the test runtime classpath of the containing project by adding the `junit-console` module (and all of its transitive dependencies) to it. Consequently, the `junit4` and `junit5` engines were always on the classpath, both for the project and for the task created by the plugin. The negative side effect is that those engines were then included in the test plan by default, even if the user didn't want either of those engines.

This commit addresses this issue by introducing a `junitPlatform` configuration that serves as a classpath container for the JUnit Platform modules required by the custom task created by the JUnit Gradle plugin. This `junitPlatform` configuration has a dependency on the `junit-launcher` (and its transitive dependencies such as the `junit-engine-api`) and a dependency on the `junit-console` (but **without** any of its transitive dependencies).

For example, executing `gradlew :junit-tests:dependencies` will result in a dependency tree similar to the following.

```
junitPlatform
+--- org.junit:junit-launcher:5.+ -> 5.0.0-SNAPSHOT
|    \--- org.junit:junit-engine-api:5.0.0-SNAPSHOT
|         +--- org.junit:junit-commons:5.0.0-SNAPSHOT
|         \--- org.opentest4j:opentest4j:1.0.0-M1
\--- org.junit:junit-console:5.+ -> 5.0.0-SNAPSHOT
```

The classpath for the custom task created by the JUnit Gradle plugin is then composed of the `junitPlatform` configuration and anything the user adds to the test run-time classpath for the project.

Consequently:

1. No TestEngines are accidentally added to the test runtime classpath of the project.
1. No TestEngines are accidentally added to the classpath of the custom task created by the JUnit Gradle plugin.
